### PR TITLE
CASMINST-6188 Import / Export of BSS Data

### PR DESCRIPTION
**Pull request by mbuchmann-hpe**
# Description

Added docs for import and export of BSS Boot Parameters and scripts to support this.  Chmod +x on the scripts.

CASMINST-6188

# Checklist

- [x] If I added any command snippets, the steps they belong to follow the prompt conventions (see [example][1]).
- [x] If I added a new directory, I also updated `.github/CODEOWNERS` with the corresponding team in [Cray-HPE][2]. (scripts directory not in CODEOWNERS file)
- [x] My commits or Pull-Request Title contain my JIRA information, or I do not have a JIRA.

 
[1]: https://github.com/Cray-HPE/docs-csm/blob/main/introduction/documentation_conventions.md#using-prompts
[2]: https://github.com/Cray-HPE/teams

### CODE REVIEW

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Export and Importing BSS Data
+
+## Prerequisites
+
+- Ensure that the `cray` command line interface (CLI) is authenticated and configured to talk to system management services.
+  - See [Configure the Cray CLI](../configure_cray_cli.md).
+- In order to use the automated procedures, the latest CSM documentation RPM must be installed on the node where the procedure is being performed.
+  - See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation).
+
+### Exporting BSS Boot Parameters
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
## Export BSS boot parameters
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+1. (`ncn-mw#`) Run the following command to create a `json` file with the boot parameters for only the Compute Nodes:
+
+   ```bash
+   xnames=`cray hsm state components list --type Node --role Compute --format json | jq -r '.[] | map(.ID) | join(",")'`
+   echo $xnames
+   cray bss bootparameters list --name $xnames --format json > cray-bss-compute-boot-parameters-dump.json
+   ```
+
+### Restore BSS Boot Parameters
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
## Restore BSS boot parameters
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   export TOKEN=$(curl -k -s -S -d grant_type=client_credentials -d client_id=admin-client -d client_secret=$(kubectl get secrets admin-client-auth -o jsonpath='{.data.client-secret}' | base64 -d) https://api-gw-service-nmn.local/keycloak/realms/shasta/protocol/openid-connect/token | jq -r '.access_token')
+   ```
+
+1. (`ncn-mw#`) Restore Boot Parameters with the following script (`cray-bss-compute-parameters-dump.json` is the name of the file created from the export command):
+
+   ```bash
+   /usr/share/doc/csm/scripts/operations/boot_script_service/bss-restore-bootparameters.sh cray-bss-compute-parameters-dump.json
+   ```
+
+### Update `ids` and `etags` after IMS import
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
## Update BSS after IMS import
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+- **After** running an IMS
+[Automated import procedure](../image_management/Exporting_and_Importing_IMS_Data.md#automated-import-procedure),
+run the following script to update the BSS boot parameters.
+The IMS import script should have generated a file containing the IMS ID and S3 `etag` mappings -- it is displayed near
+the end of the script output.
+
+1. (`ncn-mw#`)  Set up token:
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
1. (`ncn-mw#`) Set up an API token.
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+   ```bash
+   export TOKEN=$(curl -k -s -S -d grant_type=client_credentials -d client_id=admin-client -d client_secret=$(kubectl get secrets admin-client-auth -o jsonpath='{.data.client-secret}' | base64 -d) https://api-gw-service-nmn.local/keycloak/realms/shasta/protocol/openid-connect/token | jq -r '.access_token')
+   ```
+
+1. (`ncn-mw#`)  Run this script
+(`/root/ims-import-export-data/ims-id-maps-post-import-12f86451ce7c49d79e345bee42cc8586.json` is the file from the IMS import procedure):
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
1. (`ncn-mw#`) Update BSS based on the changes made during the IMS import.

   > `/root/ims-import-export-data/ims-id-maps-post-import-12f86451ce7c49d79e345bee42cc8586.json` is the file from the IMS import procedure.
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+1. (`ncn-mw#`) Restore Boot Parameters with the following script (`cray-bss-compute-parameters-dump.json` is the name of the file created from the export command):
+
+   ```bash
+   /usr/share/doc/csm/scripts/operations/boot_script_service/bss-restore-bootparameters.sh cray-bss-compute-parameters-dump.json
+   ```
+
+### Update `ids` and `etags` after IMS import
+
+- **After** running an IMS
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
**After** running an IMS
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+1. Copy the file generated by the export command to the node where the import procedure is being performed.
+
+1. (`ncn-mw#`)  Set up token:
+
+   ```bash
+   export TOKEN=$(curl -k -s -S -d grant_type=client_credentials -d client_id=admin-client -d client_secret=$(kubectl get secrets admin-client-auth -o jsonpath='{.data.client-secret}' | base64 -d) https://api-gw-service-nmn.local/keycloak/realms/shasta/protocol/openid-connect/token | jq -r '.access_token')
+   ```
+
+1. (`ncn-mw#`) Restore Boot Parameters with the following script (`cray-bss-compute-parameters-dump.json` is the name of the file created from the export command):
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
1. (`ncn-mw#`) Restore the boot parameters.

   > `cray-bss-compute-parameters-dump.json` is the name of the file created from the export procedure.
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+   xnames=`cray hsm state components list --type Node --role Compute --format json | jq -r '.[] | map(.ID) | join(",")'`
+   echo $xnames
+   cray bss bootparameters list --name $xnames --format json > cray-bss-compute-boot-parameters-dump.json
+   ```
+
+### Restore BSS Boot Parameters
+
+1. Copy the file generated by the export command to the node where the import procedure is being performed.
+
+1. (`ncn-mw#`)  Set up token:
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
1. (`ncn-mw#`) Set up an API token.
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+   ```bash
+   export TOKEN=$(curl -k -s -S -d grant_type=client_credentials -d client_id=admin-client -d client_secret=$(kubectl get secrets admin-client-auth -o jsonpath='{.data.client-secret}' | base64 -d) https://api-gw-service-nmn.local/keycloak/realms/shasta/protocol/openid-connect/token | jq -r '.access_token')
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
   export TOKEN=$(curl -k -s -S -d grant_type=client_credentials -d client_id=admin-client \
       -d client_secret=$(kubectl get secrets admin-client-auth -o jsonpath='{.data.client-secret}' | base64 -d) \
       https://api-gw-service-nmn.local/keycloak/realms/shasta/protocol/openid-connect/token | jq -r '.access_token')
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+- **After** running an IMS
+[Automated import procedure](../image_management/Exporting_and_Importing_IMS_Data.md#automated-import-procedure),
+run the following script to update the BSS boot parameters.
+The IMS import script should have generated a file containing the IMS ID and S3 `etag` mappings -- it is displayed near
+the end of the script output.
+
+1. (`ncn-mw#`)  Set up token:
+
+   ```bash
+   export TOKEN=$(curl -k -s -S -d grant_type=client_credentials -d client_id=admin-client -d client_secret=$(kubectl get secrets admin-client-auth -o jsonpath='{.data.client-secret}' | base64 -d) https://api-gw-service-nmn.local/keycloak/realms/shasta/protocol/openid-connect/token | jq -r '.access_token')
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
   export TOKEN=$(curl -k -s -S -d grant_type=client_credentials -d client_id=admin-client \
       -d client_secret=$(kubectl get secrets admin-client-auth -o jsonpath='{.data.client-secret}' | base64 -d) \
       https://api-gw-service-nmn.local/keycloak/realms/shasta/protocol/openid-connect/token | jq -r '.access_token')
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+## Prerequisites
+
+- Ensure that the `cray` command line interface (CLI) is authenticated and configured to talk to system management services.
+  - See [Configure the Cray CLI](../configure_cray_cli.md).
+- In order to use the automated procedures, the latest CSM documentation RPM must be installed on the node where the procedure is being performed.
+  - See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation).
+
+### Exporting BSS Boot Parameters
+
+1. (`ncn-mw#`) Run the following command to create a `json` file with the boot parameters:
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
1. (`ncn-mw#`) Create a JSON file with the boot parameters.
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+   ```bash
+   cray bss bootparameters list --format json > cray-bss-boot-parameters-dump.json
+   ```
+
+1. (`ncn-mw#`) Run the following command to create a `json` file with the boot parameters for only the Compute Nodes:
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
1. (`ncn-mw#`) Create a JSON file with the boot parameters for only the compute nodes.
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 13]
```diff
@@ -0,0 +1,61 @@
+# Export and Importing BSS Data
+
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md:7]:
```suggestion

- [Prerequisites](#prerequisites)
- [Export BSS boot parameters](#export-bss-boot-parameters)
- [Import BSS boot parameters](#import-bss-boot-parameters)
- [Update BSS after IMS import](#update-bss-after-ims-import)

```

File: operations/image_management/Exporting_and_Importing_IMS_Data.md [hunk 14]
```diff
@@ -220,7 +220,9 @@ Make a note of the filename containing the IMS ID and S3 etag mappings -- it is
 It is important to retain this mapping file for later reference. In particular, its contents will
 be needed in order to update data in other services, such as BOS and BSS. Save this file in a safe location.
 See [Automated BOS data import](../boot_orchestration/Exporting_and_Importing_BOS_Data.md#automated-bos-data-import)
-for more information on when and how this file is used during BOS data imports.
+and
+[Restore BSS boot parameters](../boot_orchestration/Exporting_and_Importing_BSS_Data.md#restore-bss-boot-parameters)
```
Comment (mharding-hpe) [inline operations/image_management/Exporting_and_Importing_IMS_Data.md]:
```suggestion
[Update BSS after IMS import](../boot_orchestration/Exporting_and_Importing_BSS_Data.md#update-bss-after-ims-import)
```

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md:1]: This new file should also be linked to from `operations/README.md` (in both the boot orchestration section and the backup/restore section).

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md [hunk 15]
```diff
@@ -0,0 +1,61 @@
+# Export and Importing BSS Data
```
Comment (mharding-hpe) [inline operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md]:
```suggestion
# Exporting and Importing BSS Data
```

**Review by mharding-hpe — CHANGES_REQUESTED** [review]
Mostly linting nits. Looks good otherwise.

After the linting updates are done and the markdown checks are satisfied, this will also need backports for the 1.3 branch, the 1.5 branch, and the 1.6 branch.

File: operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md
Comment (mbuchmann-hpe) [reply operations/boot_orchestration/Exporting_and_Importing_BSS_Data.md:1]: Added

**Review by mharding-hpe — APPROVED** [review]
Also need this backported to 1.3, 1.5, and 1.6

**Review by rustydb — APPROVED** [review]

**Pull request merged**